Cardano shelley update 1/3

gabrielKerekes opened the pull request:
Cardano Shelley update issue: #948 

As promised (#1048) we are splitting the Cardano Shelley update into (at least) 3 PRs. This is the first of them.

This PR contains minimal changes with which Trezor can work on the new Shelley fork. It enables sending funds from one Byron address to another. New addresses are not yet added. Neither are certificates and withdrawals. 

**What's changed:**
- Previous transactions (input sources) no longer need to be requested/received since Fee is included directly in the transaction.
- TTL parameter has been added to transaction - this is the slot number until which the transaction can be submited.
- Transaction format has changed a bit.
- The biggest changes are in the `sign_tx.py` file. The logic around fee calculation has been removed and it has also been refactored quite a lot - e.g. Transaction class has been removed since it now mostly contained only functions anyways.
- Support for CBOR NULL had to be added to support empty transaction metadata.
- Protocol magic is included in Byron addresses on testnets - this has always been a part of the spec, but it hasn't been needed so far. We now also verify that outputs are from the network which the protocol magic represents.

**What can still change in the following days (regarding this PR):**
- ~~We still don't have specs/information about witnesses for Byron inputs from IOHK. This will probably require changes to the way the witnesses are built for Byron inputs. There's also a todo there.~~ This has already been added.
- ~~Protocol magic in Byron addresses on Testnets - this functionality is drafted, but also needs to be confirmed by IOHK.~~ This has already been added.
- Maybe some other minor changes.

[Here's a gist](https://gist.github.com/gabrielKerekes/b8b929614e11bea2ff8af6affe4caf0b) with a sample transaction input file for trezorctl (although tests also provide some samples).
[Here's a gist](https://gist.github.com/gabrielKerekes/85dd1c0bf1059e4c2627fc2ee719cb50) with a description of the Shelley transaction format (including certificates and withdrawals).
[Here's](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/executable-spec/cddl-files/shelley.cddl) the official CDDL spec for Shelley.

**This hasn't been tested on the latest testnet yet, since Byron addresses are not yet fully supported. Support should be added in the following days. When this happens, we will test it and let you know of the status** (Some changes may still need to be made regarding this).

prusnak commented:
For the record: IOHK wants to do a hard-fork on the 29th of July.

MichalPetro commented:
@matejcik @tsusanka guys are you going to review this? 2 more PRs will be coming in the following week as we divided the functionalities in 3 PRs.

tsusanka commented:
@MichalPetro we will look into this ASAP, tomorrow most likely.

----

Regarding the PR schedule: we would like to "freeze" (branch of code into a separate branch and forward firmwares to QA) the upcoming release on July 24th. That gives us 10 working days to include the necessary changes.

Although I believe this was mentioned somewhere please confirm that to support sending funds we need to include this PR and one more you are about to open. Please open this PR asap so we can review it and include it till July 24th. Apologies for the sudden rush, I haven't realized the hard-fork date is so close.

As to the third PR (which will introduce staking support, correct?) we need to discuss this internally in the Product team whether we want such change in firmware. In another words, it depends on the number of needed changes. We usually support only a subset of functionalities for altcoins because it is not sustainable to have advanced functions for every altcoin. I am basically trying to manage expectations here - while we definitely want to include PR 1 and 2 I can not guarantee the same for PR number 3. Feel free to file a draft PR again to decide this before you fine-tune it.

MichalPetro commented:
@tsusanka we will add the 2nd PR today or tomorrow. Also, we probably need to do the small fix in the 1st (this) PR because some specs changed, we are currently assessing the impact.

The 3rd PR will be smaller than the others. We will be not introducing staking for stake pools, just staking delegation to the stake pools.

tsusanka commented:
Great, thank you.

> The 3rd PR will be smaller than the others. We will be not introducing staking for stake pools, just staking delegation to the stake pools.

Ok ok. But it is something we might include in the next release if things go south, correct? It is not _that_ essential as the other two.

MichalPetro commented:
Yes definitely the first 2 PRs are critical. Lets make sure those are included. Lets see what we can do about the 3rd.

gabrielKerekes commented:
I've updated the code with:
- correct mainnet protocol magic
- correct transaction witnesses
- correct way of including protocol magic in Byron on addresses on Testnet
- some type hinting

Hopefully nothing else will change. I've kept the fixup commits (which is failing the checks), so it's easy to see what changed. I can rebase them or squash them or whatever if you wish.

matejcik commented on core/src/apps/cardano/protocol_magics.py:
```diff
@@ -0,0 +1,12 @@
+class ProtocolMagics:
```
any reason this is a class as opposed to either a static dict (doesn't seem to be too big), or just a function (if RAM usage is a concern)?

gabrielKerekes replied:
We've made it into a class for convenience.

Is something like this what you had in mind?
```
PROTOCOL_MAGICS = {"mainnet": 764824073, "testnet": 42}


def protocol_magic_to_ui_string(protocol_magic: int) -> str:
    ui_string_map = {
        PROTOCOL_MAGICS["mainnet"]: "Mainnet",
        PROTOCOL_MAGICS["testnet"]: "Testnet",
    }

    return ui_string_map.get(protocol_magic, "Unknown")
```

gabrielKerekes replied:
Updated in this [commit](https://github.com/trezor/trezor-firmware/pull/1096/commits/71f2191b6f0a3c4ecea9d0ded1a2f62b814283a4).

matejcik replied:
more like:
```
MAINNET = 764824073
TESTNET = 42

NAMES = {
	MAINNET: "Mainnet",
	TESTNET: "Testnet",
}

def to_ui_string(value: int) -> str:
	return NAMES.get(value, "Unknown")
```

for ease of use. then you can refer to `protocol_magics.MAINNET` for the constant, `protocol_magics.to_ui_string(xyz)` for the query

matejcik replied:
basically you're using a class to do a module's job, just move everything you had before to top-level

gabrielKerekes replied:
Ok, I got it. Thanks! Updated [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/f4a1990e9e8dbf1721314eba69aed8e4d0e535e4).

matejcik commented:
> I've kept the fixup commits (which is failing the checks), so it's easy to see what changed. I can rebase them or squash them or whatever if you wish.

Thanks.
Our review flow is as follows:
- once you request reviews, only add fixup commits (you are doing that already, so :+1:)
- respond to individual review comments that this is done, ideally with a link to commit
- the reviewer marks comments resolved
- when you get all approvals, autosquash the fixup commits and force-push

matejcik commented:
please regenerate UI test fixtures, by running the test suite with `--ui=record`

tsusanka commented:
> please regenerate UI test fixtures, by running the test suite with --ui=record

Or simply use `pipenv run make -C core test_emu_ui_record`, see https://docs.trezor.io/trezor-firmware/tests/ui-tests.html#updating-fixtures-recording.

gabrielKerekes commented on common/protob/messages-cardano.proto:
```diff
@@ -14,8 +14,9 @@ import "messages-common.proto";
-    optional bool show_display = 2; // optionally prompt for confirmation on trezor display
+    repeated uint32 address_n = 1;      // BIP-32-style path to derive the key from master node
+    optional bool show_display = 2;     // optionally prompt for confirmation on trezor display
+    optional uint32 protocol_magic = 3; // network's protocol magic - needed for Byron addresses on testnets
```
This was a requirement we received from IOHK as part of the Shelley update - to also support Byron addresses on testnets. I hope this isn't too much of a problem.

gabrielKerekes commented on core/src/apps/cardano/address.py:
```diff
@@ -1,54 +1,104 @@
+    return _encode_address_raw(address_data_encoded) == address
+
+
+def _get_address_data(address: str) -> bytes:
```
It's returning encoded data because the `is_safe_output_address` uses this data cbor encoded to try and recreate the address. The `matches_with_procol_magic` decodes this data further.

gabrielKerekes replied:
`raise INVALID_ADDRESS` added in this [commit](https://github.com/trezor/trezor-firmware/pull/1096/commits/6ca9b50238a76e0aedecc6f8af3a001045ddb606).

gabrielKerekes replied:
Hm, I just realised I should probably rename the function to `_get_encoded_address_data` then. I will do that soon.

matejcik replied:
right, I got confused because the address data is encoded twice, and in both cases the result is a 2-tuple

but yes, please rename the function

gabrielKerekes commented on core/src/apps/cardano/address.py:
```diff
@@ -74,17 +124,13 @@ def validate_full_path(path: list) -> bool:
-    data = cbor.encode(data)
-    data = hashlib.sha3_256(data).digest()
-    res = hashlib.blake2b(data=data, outlen=28).digest()
+def _address_hash(data: list) -> bytes:
```
Yes, I am not sure what the right solution here would be though.

gabrielKerekes replied:
I hope to get the types in order after the Shelley update.

gabrielKerekes commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,238 +4,195 @@
+        keychain, msg.inputs, tx_hash, msg.protocol_magic
+    )
+    # byron witnesses have the key 2 in Shelley
+    witnesses = {2: _detupleize(witnesses_for_cbor)}
```
This was done purely so that the types can be set correctly and therefore readability can be improved. 

We also thought about implementing it in CBOR although then there could be an ambiguity in decoding arrays - do we decode as list or a tuple?

matejcik replied:
Just decode as array always. I'd file this under "API convenience": CBOR only recognizes arrays (right?), and so only returns arrays. But for usage convenience it will transparently encode tuples as arrays as well, because that's how we use it

gabrielKerekes replied:
Done [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/0ead85323b19db525f48336936246602f066f8eb). I'll change the commit message later, if we're not squashing to one commit.

gabrielKerekes commented on core/src/apps/cardano/address.py:
```diff
@@ -1,54 +1,104 @@
+    if len(address_attributes) == 0:
+        return protocol_magic == ProtocolMagics.MAINNET
+
+    if 2 not in address_attributes:
```
Done [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/6ca9b50238a76e0aedecc6f8af3a001045ddb606).

gabrielKerekes commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,238 +4,195 @@
-        transaction.inputs,
-        transaction.outputs,
-    )
+    await _show_tx(ctx, keychain, msg)
```
Done [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/ce30016fe552cd4991eb34eddccab83561cb39d1).

I also had to extract output validation so it happens before showing the tx so it doesn't crash unexpectedly. The separation is a good thing anyways, I think.

gabrielKerekes commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -4,238 +4,195 @@
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.change_coins[index]]
+def _build_inputs(inputs: List[CardanoTxInputType]) -> List[Tuple[bytes, int]]:
+    result = []
```
Nice. Done [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/ce30016fe552cd4991eb34eddccab83561cb39d1).

gabrielKerekes commented:
I've updated the code with the suggestions. I've also updated the fixtures json [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/83b9ae3e27a7bd71ca6ff963529eeca019454465).

gabrielKerekes commented:
I had to update transaction signing due to Byron witnesses spec [update](https://github.com/input-output-hk/cardano-ledger-specs/pull/1670). [My commit](https://github.com/trezor/trezor-firmware/pull/1096/commits/6e267d591f5d2bbc33b5e0075c540d9e6591c95a). I am however unable to test this at the moment since they also have to release their version of cardano-node.

I suggest we get the code in order and I will test everything once more when it's ready. I expect this to be on monday.

matejcik commented on core/src/apps/cardano/layout/__init__.py:
```diff
@@ -6,6 +6,7 @@
 from trezor.ui.text import Text
 from trezor.utils import chunks
 
+from apps.cardano.protocol_magics import protocol_magic_to_ui_string
```
a thing i forgot previously:

please use relative imports for all `apps.cardano` imports, in all files under `apps/cardano`

i.e.:

```
from apps.common.confirm import (...)

from ..protocol_magics import to_ui_string
```

gabrielKerekes replied:
Done in [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/10305274dbc6f66e0e7007c0a7ecbc0fddbc1372) and [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/7a2d04e20370f03fc47dd24f7e02286866b1ff0a).

matejcik commented on core/src/apps/cardano/address.py:
```diff
@@ -1,54 +1,104 @@
     """
     Determines whether it is safe to include the address as-is as
     a tx output, preventing unintended side effects (e.g. CBOR injection)
     """
```
ISTM now that you're not using `Raw` anymore, there is no CBOR injection. I am not sure what this check is accomplishing exactly.

In fact the structure of the whole file is unclear. There is `_encode_address_raw`, and `_get_address_data` seems to be its counterpart -- but in that case, it should perform the CRC check, and maybe even validate that the CBOR tag is present? I'm not sure how important that is though.
Also the names should make it clear that these two functions perform the reverse operations. Something like `_wrap_address_raw` and `_unwrap_address_raw`?

And then I'd rename `matches_with_protocol_magic` to `validate_address(address: str, protocol_magic: int)` and make it raise either `INVALID_ADDRESS` if the format is wrong, or "Network mismatch" if the protocol magic is wrong -- so that in `sign_tx` you can just call `validate_address()` without the ifs

matejcik replied:
While you're at this, please remove `Raw` from the CBOR encoder. It's dangerous and now that we don't need it, there is no reason for it to live in the code.

gabrielKerekes replied:
We are still using `Raw` so the check is still valid. I agree that the file is unclear, but it's also given by the fact that the Byron address format is complicated. I'm trying to refactor it with your notes in mind.

matejcik replied:
i don't see any instances of `Raw` in the code. Are Shelley addresses using `Raw`?

gabrielKerekes replied:
Hm, I see. I was a bit confused. Indeed, output addresses are encoded as bytes now. Shelley addresses are also not using `Raw`. So this check can be removed - I'll confirm this with the team.

gabrielKerekes replied:
Done [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/10305274dbc6f66e0e7007c0a7ecbc0fddbc1372). CBOR injection check has been removed. We now need to deconstruct the address only to get to the protocol magic. CRC doesn't need to be checked since that is done by the node.

I've also removed `Raw` [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/0ead85323b19db525f48336936246602f066f8eb) from CBOR. Unfortunately I can't remove the 24 Tag since it's still used when building the address from path.

matejcik replied:
The CRC check should be there anyway, to prevent user error. We shouldn't ask the user to confirm address strings which we can determine are invalid.
Unless there is a valid usecase for generating txes to invalid addresses, of course .)

gabrielKerekes replied:
Added [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/ea4a4b36c687f32f9d3385aef083bccfe7631e78). I also added a test case for that.

gabrielKerekes replied:
Oh, I need to generate UI fixtures again.

matejcik commented:
ah, I think I said this wrong

> * once you request reviews, only add fixup commits (you are doing that already, so +1)

You *can* add new commits. The "only fixup" thing is intended as "no rebasing", so modifications to existing commits should be fixups. But if the changes don't cleanly match some existing commits, a new commit is a better variant.
(you can also fully revert commits via `git revert`, but then you need to make sure during autosquash that the commit/revert pairs are removed)

gabrielKerekes commented:
Yes, that makes sense. I wasn't sure how to go about that so I kept on "fixing-up". I'll know better next time.

matejcik commented:
yeah, sorry about that :roll_eyes:

gabrielKerekes commented:
The second PR needs to be rebased onto this one - I've partially done this already, but I'll probably have to do it again. Then I'll clean it up, test it (also against the testnet) and it should be ready for review - hopefully until tomorrow afternoon.

And yes, we can probably do something like that. Should I squash and force push the fixups?

matejcik commented:
> And yes, we can probably do something like that. Should I squash and force push the fixups?

not yet, please wait for approval from @tsusanka

tsusanka commented:
Thanks, great job! Feel free to squash the commits now and force push. Then change this PR to be against the `cardano-shelley` branch I've just created.

gabrielKerekes commented:
Awesome! Should I squash all the comits into one, or just the fixups?

matejcik commented:
up to you honestly. ISTM that via the fixup confusion the original commits don't make sense separately anyway, so feel free to squash everything together.
alternately you can rename some commits and keep some semblance of history

gabrielKerekes commented:
I've opened the second PR -> #1112
